src/scheduler/fifo.c

```diff
@@ -1727,11 +1727,13 @@ run_update_resresv(status *policy, int pbs_sd, server_info *sinfo,
 				int j;
 				update_node_on_run(ns[i], rr, &old_state);
 				if (ns[i]->ninfo->np_arr != NULL) {
-					for (j = 0; ns[i]->ninfo->np_arr[j] != NULL; j++) {
-						modify_resource_list(ns[i]->ninfo->np_arr[j]->res, ns[i]->resreq, SCHD_INCR);
+					node_partition **npar = ns[i]->ninfo->np_arr;
+					for (j = 0; npar[j] != NULL; j++) {
+						modify_resource_list(npar[j]->res, ns[i]->resreq, SCHD_INCR);
 						if (!ns[i]->ninfo->is_free)
-							ns[i]->ninfo->np_arr[j]->free_nodes--;
+							npar[j]->free_nodes--;
 						sort_nodepart = 1;
+						update_buckets_for_node_array(npar[j]->bkts, npar[j]->ninfo_arr);
```

Comment:
This loops over each node, and you are updating the node partition's buckets for an array.  This means if we will likely update the same node partition multiple times since we run a job on nodes from a single node partition.  You should rip the guts out of update_buckeys_for_node_array() and make it work for a single node.  You can call the new function in both the old function and here.